lib/mock/mock-agent.js

```diff
@@ -0,0 +1,141 @@
+'use strict'
+
+const { kOptions, kClients, kFactory } = require('../core/symbols')
+const Agent = require('../agent')
+const {
+  kAgent,
+  kMockAgentSet,
+  kMockAgentGet,
+  kDispatches,
+  kIsMockActive,
+  kNetConnect,
+  kGetNetConnect
+} = require('./mock-symbols')
+const MockClient = require('./mock-client')
+const MockPool = require('./mock-pool')
+const { matchValue } = require('./mock-utils')
+const { InvalidArgumentError } = require('../core/errors')
+const Dispatcher = require('../dispatcher')
+
+/* istanbul ignore next: difficult to make a test case for */
+const WeakRef = global.WeakRef || class CompatWeakRef {
+  constructor (value) {
+    console.log('----------------')
+    this.value = value
+  }
+
+  deref () {
+    return this.value.connected === 0 && this.value.size === 0
+      ? undefined
+      : this.value
+  }
+}
+
+class MockAgent extends Dispatcher {
+  constructor (opts) {
+    super(opts)
+
+    this[kNetConnect] = true
+    this[kIsMockActive] = true
+
+    // Instantiate Agent and encapsulate
+    if ((opts && opts.agent && typeof opts.agent.dispatch !== 'function')) {
+      throw new InvalidArgumentError('Argument opts.agent must implement Agent')
+    }
+    const agent = opts && opts.agent ? opts.agent : new Agent(opts)
+    this[kAgent] = agent
+
+    this[kClients] = agent[kClients]
+    this[kOptions] = agent[kOptions]
```

Comment:
This is a little dangerous. But I'm not familiar enough with this to suggest anything better. I could easily change the internals in the future and accidentally break mocking support. Hope we have test that can catch that.

Reply:
Understood and yep, we have a test to catch this. I'll try and come up with sometime better that doesn't reference `kClients`, `kOptions` or `kFactory`.

Reply:
@ronag this is the exact reason why we need mocking support inside Undici.

Reply:
FYI both, I have moved `kOptions` and `kFactory` back into the Agent as these were not core to the mocking functionality, so only `kClients` is referenced (and backed up with tests so if this is changed, the mocks will soon know and should be a simple fix to resolve)

Reply:
@ronag is this solution okay with you? I've added a specific unit test to immediately catch if `kClients` changes. The main problem I found with having a separate `kClients` symbol in mock agent was that we would lose agent encapsulation and need to add additional code for close and dispatch (in other words, it's doable but with some drawbacks). Let me know what you think!